bgpy/simulation_engine/policies/aspa/aspa.py

```diff
@@ -40,104 +41,144 @@ def _valid_ann(self, ann: "Ann", from_rel: Relationships) -> bool:  # type: igno
         else:
             raise NotImplementedError("Should never reach here")
 
-    def _upstream_check(self, ann: "Ann", from_rel: Relationships) -> bool:
+    def _upstream_check(self, ann: "Ann", from_rel: "Relationships") -> bool:
         """ASPA upstream check"""
 
         # Upstream check
         if len(ann.as_path) == 1:
             return super()._valid_ann(ann, from_rel)
         else:
-            reversed_path = ann.as_path[::-1]  # type: ignore
             # For every adopting ASPA AS in the path,
             # The next ASN in the path must be in their providers list
-            for i in range(len(reversed_path) - 1):
-                if not self._provider_check(reversed_path[i], reversed_path[i + 1]):
-                    return False
 
-        return super()._valid_ann(ann, from_rel)  # type: ignore
+            # 4. If max_up_ramp < N, the procedure halts with the outcome "Invalid".
+            if self.get_max_up_ramp_length(ann) < len(ann.as_path):
+                return False
+
+            # No need to actually verify the min_up_ramp as we accept Unknown announcements
+            # 5. If min_up_ramp < N, the procedure halts with the outcome "Unknown".
+            if self.get_min_up_ramp_length(ann) < len(ann.as_path):
+                # Valid as we do not invalidate Unknown announcements
+                return super()._valid_ann(ann, from_rel)
+
+        # Actually ASPA-valid
+        return super()._valid_ann(ann, from_rel)
+
+    def get_max_up_ramp_length(self, ann: "Ann") -> int:
+        """Upstream ramp length function, responsible for the qualification into either Invalid or Unknown/Valid"""
+        reversed_path = ann.as_path[::-1]
+        uppercase_i = iterate_check_func_over_asn_path(
+            reversed_path, self._provider_or_peer_check
+        )
+
+        # If there is no such I, the maximum up-ramp length is set equal to the AS_PATH length N
+        if uppercase_i == len(reversed_path) - 1:
+            max_up_ramp = len(reversed_path)
+        else:
+            max_up_ramp = uppercase_i
+
+        return max_up_ramp
+
+    def get_min_up_ramp_length(self, ann: "Ann") -> int:
+        """Upstream ramp length function, responsible for the qualification into either Unknown or Valid"""
+        reversed_path = ann.as_path[::-1]
+        uppercase_i = iterate_check_func_over_asn_path(
+            reversed_path, self._provider_check
+        )
+        # If there is no such I, the AS_PATH consists of only "Provider+" pairs;
+        # so the minimum up-ramp length is set equal to the AS_PATH length N.
+        if uppercase_i == len(reversed_path) - 1:
+            min_up_ramp = len(reversed_path)
+        else:
+            min_up_ramp = uppercase_i
```

Comment:
I think I'm missing something here, not sure why it has to modified on line 91 but not line 93